### PR TITLE
Add documentation for build-time code analysis

### PR DESCRIPTION
**Pull request by stjj89**
These docs explain how to write, register, and configure
build-time code analysis checks.

### CODE REVIEW

File: go/checks.rst [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+        srcs = [
+            "check_dom.go",
+            "dom_utils.go",
+        ],
+        importpath = "unsafedom",
+        deps = ["@io_bazel_rules_go//go/tools/analysis:analysis"],
+        visibility = ["//visibility:public"],
+    )
+
+You must define a ``go_checker`` target whose ``deps`` attribute contains all
```
Comment (jayconrod) [inline go/checks.rst]: A short explanation of what this rule builds would be useful.

File: go/checks.rst [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        deps = [
+            ":mycheck",
+            ":unsafedom",
+            "@javascript_checks//:loopclosure", # we can import checks from a remote repo
+        ],
+        visibility = ["//visibility:public"],
+    )
+
+Finally, the ``go_checker`` target must be passed to ``go_register_toolchains``
+in your ``WORKSPACE`` file.
```
Comment (jayconrod) [inline go/checks.rst]: Might be worth noting that checks apply to all libraries built within the workspace, including those in external repos. Checks won't apply to libraries in the current repo imported into other workspaces though.

File: go/checks.rst [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+.. code:: bzl
+
+    go_checker(
+        name = "my_checker",
+        deps = [
+            ":mycheck",
+            ":unsafedom",
+            "@javascript_checks//:loopclosure",
+        ],
+        config = ":config.json"
```
Comment (jayconrod) [inline go/checks.rst]: Nit: Don't include the leading `:` in the filename. It doesn't make any difference to Bazel, but it helps distinguish rules from static files.

File: go/checks.rst [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+| A unique name for this rule.                                                                     |
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`deps`              | :type:`label_list`          | :value:`None`                         |
++----------------------------+-----------------------------+---------------------------------------+
+| List of Go libraries that will be linked to the generated checker binary.                        |
+| These libraries must call analysis.Register to ensure that the analyses they implement are       |
+| called by the checker binary.                                                                    |
+| These libraries must be `go_tool_library` rules to avoid bootstrapping problems.                 |
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`config`             | :type:`label`          | :value:`None`                             |
```
Comment (jayconrod) [inline go/checks.rst]: This isn't rendering correctly. I think there's some misalignment.

File: go/checks.rst [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+| **Name**                   | **Type**                    | **Default value**                     |
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`name`              | :type:`string`              | |mandatory|                           |
++----------------------------+-----------------------------+---------------------------------------+
+| A unique name for this rule.                                                                     |
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`deps`              | :type:`label_list`          | :value:`None`                         |
++----------------------------+-----------------------------+---------------------------------------+
+| List of Go libraries that will be linked to the generated checker binary.                        |
+| These libraries must call analysis.Register to ensure that the analyses they implement are       |
```
Comment (jayconrod) [inline go/checks.rst]: Format `analysis.Register` appear as code using double backticks.

File: go/checks.rst [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+| called by the checker binary.                                                                    |
+| These libraries must be `go_tool_library` rules to avoid bootstrapping problems.                 |
```
Comment (jayconrod) [inline go/checks.rst]: `go_tool_binary` should either be formatted as code or link to the rule; ideally both, but I don't think ReST can do that.

File: go/checks.rst [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+            ":mycheck",
+            ":othercheck",
+            "@javascript_checks//:unsafedom", # we can import checks from a remote repo
+        ],
+        config = ":config.json"
+        visibility = ["//visibility:public"],
+    )
+
+
+go_tool_library
```
Comment (jayconrod) [inline go/checks.rst]: This documentation should probably go in core.rst. It's not strictly part of the checking framework.

File: go/checks.rst [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+| :param:`name`              | :type:`string`              | |mandatory|                           |
++----------------------------+-----------------------------+---------------------------------------+
+| A unique name for this rule.                                                                     |
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`srcs`              | :type:`label_list`          | :value:`None`                         |
++----------------------------+-----------------------------+---------------------------------------+
+| The list of Go source files that are compiled to create the package.                             |
+| Only :value:`.go` files are permitted, unless the cgo attribute is set, in which case the        |
+| following file types are permitted: :value:`.go, .c, .s, .S .h`.                                 |
+| The files may contain Go-style `build constraints`_.                                             |
```
Comment (jayconrod) [inline go/checks.rst]: Need a destination for this link.

File: go/checks.rst [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
++----------------------------+-----------------------------+---------------------------------------+
+| List of Go libraries this library imports directly.                                              |
+| These may be go_library rules or compatible rules with the GoLibrary_ provider.                  |
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`embed`             | :type:`label_list`          | :value:`None`                         |
++----------------------------+-----------------------------+---------------------------------------+
+| List of Go libraries this test library directly.                                                 |
+| These may be go_library rules or compatible rules with the GoLibrary_ provider.                  |
+| These can provide both :param:`srcs` and :param:`deps` to this library.                          |
+| See Embedding_ for more information about how and when to use this.                              |
```
Comment (jayconrod) [inline go/checks.rst]: Need a destination for this link.

File: go/checks.rst [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`data`              | :type:`label_list`          | :value:`None`                         |
++----------------------------+-----------------------------+---------------------------------------+
+| The list of files needed by this rule at runtime. Targets named in the data attribute will       |
+| appear in the *.runfiles area of this rule, if it has one. This may include data files needed    |
+| by the binary, or other programs needed by it. See `data dependencies`_ for more information     |
```
Comment (jayconrod) [inline go/checks.rst]: Need a destination for this link.

File: go/checks.rst [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+| about how to depend on and use data files.                                                       |
++----------------------------+-----------------------------+---------------------------------------+
+
+Example
+^^^^^^^
+
+.. code:: bzl
+
+  go_tool_library(
```
Comment (jayconrod) [inline go/checks.rst]: Alignment seems wrong here.

File: go/providers.rst [hunk 12]
```diff
@@ -303,3 +303,16 @@ each package.
 | * ``data``: list of data ``File``s.                                                              |
 +--------------------------------+-----------------------------------------------------------------+
 
+GoChecker
```
Comment (jayconrod) [inline go/providers.rst]: Do we actually need this provider? It doesn't seem like it provides anything useful. The executable is already going to be in `DefaultInfo`.

Comment (stjj89) [reply go/providers.rst]: Good point. Removed.

File: go/checks.rst (hunk 11 above)
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst (hunk 10 above)
Comment (stjj89) [reply go/checks.rst]: Moved to core.rst. No longer relevant.

File: go/checks.rst (hunk 9 above)
Comment (stjj89) [reply go/checks.rst]: Moved to core.rst. No longer relevant.

File: go/checks.rst (hunk 8 above)
Comment (stjj89) [reply go/checks.rst]: Moved to core.rst. No longer relevant.

File: go/checks.rst (hunk 7 above)
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst (hunk 6 above)
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst (hunk 5 above)
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst (hunk 4 above)
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst (hunk 3 above)
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst (hunk 2 above)
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst (hunk 1 above)
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+    func init() {
+      analysis.Register(Analysis)
+    }
+
+    func run(p *analysis.Package) (*analysis.Result, error) {
+      // Analysis logic goes here.
+    }
+
+Each check must be built as a ``go_tool_library`` target. This rule is
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/checks.rst]: Done. Your suggested writeup is much better.

File: go/checks.rst [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+    import "github.com/bazelbuild/rules_go/go/tools/analysis"
+
+    var Analysis = &analysis.Analysis{Name: "mycheck", Run: run}
+
+    func init() {
+      analysis.Register(Analysis)
+    }
+
+    func run(p *analysis.Package) (*analysis.Result, error) {
+      // Analysis logic goes here.
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/checks.rst]: Done. I made this a check that looks for an import of package unsafe.

File: go/checks.rst [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+
+.. role:: param(kbd)
+.. role:: type(emphasis)
+.. role:: value(code)
+.. |mandatory| replace:: **mandatory value**
+
+rules_go allows you to define custom source-code-level checks that are executed
+alongside the Go compiler. These checks can be used to catch bugs early in the
+development process.
+
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+.. _GoLibrary: providers.rst#GoLibrary
+.. _GoSource: providers.rst#GoSource
+.. _GoArchive: providers.rst#GoArchive
+.. _GoChecker: providers.rst#GoChecker
+
+.. role:: param(kbd)
+.. role:: type(emphasis)
+.. role:: value(code)
+.. |mandatory| replace:: **mandatory value**
+
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/checks.rst]: Done. I have added a warning for this.

File: go/checks.rst [hunk 17]
```diff
@@ -0,0 +1,302 @@
+Go build-time code analysis
+===========================
+
+.. _go_library: core.rst#go_library
+.. _analysis: tools/analysis/analysis.go
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/checks.rst]: I tried [searching](https://godoc.org/?q=github.com%2Fbazelbuild%2Frules_go%2Fgo%2Ftools%2Fanalysis) for the analysis package on godoc.org but couldn't find it. Does the package need to make it to the master branch of `rules_go` before it gets parsed by Godoc?

File: go/checks.rst [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+The following configuration file configures the checks named ``mycheck``,
+``unsafedom``, and ``loopclosure``.
+
+.. code:: json
+
+    {
+      "mycheck": {
+        "severity": "ERROR",
+        "whitelist": {
+          "src/foo.go": "manually verified that behavior is working-as-intended"
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/checks.rst]: Done.

File: go/checks.rst [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+      // Analysis logic goes here.
+    }
+
+Each check must be built as a ``go_tool_library`` target. This rule is
+identical to `go_library`_ but is defined separately to avoid bootstrapping
+problems.
+
+.. code:: bzl
+
+    go_tool_library(
```
[the remark this answers is not in the document]
Comment (stjj89) [reply go/checks.rst]: Done.

**Comment by jayconrod** [comment]
Looks really good. Thanks!

By the way, I asked John on the Bazel team if there's a way we can get `go_checker` to depend on regular `go_library` rules *and* be part of the toolchain. I think that would be a much nicer API. It doesn't seem like this is possible today, but there's a proposal out for something like this. I've filed bazelbuild/bazel#5345 to track this.

**Pull request merged**